After provisioning from the catalog I cannot login to the console with my OpenShift user 0.23.0

**Describe the bug**
Using EnMasse 0.23.0 when provisioning from the catalog everything looks to have worked but I cannot login to the EnMasse console.

Digging further I see my OpenShift user has not been added the keycloak realm.

Looking at the keycloak controller I see the following exception

```

io.enmasse.user.model.v1.UserValidationFailedException: io.enmasse.user.model.v1.UserValidationFailedException: Invalid resource name 'mine.admin@example.com', must match ^[a-z]+([a-z0-9\-]*[a-z0-9]+\|[a-z0-9]*)\.[a-z0-9]+([a-z0-9\-]*[a-z0-9]+\|[a-z0-9]*)$
--
  | at io.enmasse.user.model.v1.User.validate(User.java:61)
  | at io.enmasse.user.keycloak.KeycloakUserApi.createUser(KeycloakUserApi.java:155)
  | at io.enmasse.keycloak.controller.KeycloakManager.onUpdate(KeycloakManager.java:114)
  | at io.enmasse.k8s.api.ResourceChecker.doWork(ResourceChecker.java:48)
  | at io.enmasse.k8s.api.ResourceChecker.run(ResourceChecker.java:40)
  | at java.lang.Thread.run(Thread.java:748)
  | Caused by: io.enmasse.user.model.v1.UserValidationFailedException: Invalid resource name 'mine.admin@example.com', must match ^[a-z]+([a-z0-9\-]*[a-z0-9]+\|[a-z0-9]*)\.[a-z0-9]+([a-z0-9\-]*[a-z0-9]+\|[a-z0-9]*)$
  | at io.enmasse.user.model.v1.UserMetadata.validate(UserMetadata.java:70)
  | at io.enmasse.user.model.v1.User.validate(User.java:58)

```

**To Reproduce**
Steps to reproduce the behavior:
1. Go to 'catalog'
2. Click on 'EnMasse standard and provision'
3. Once provisioned try to login it will fail
4. look at logs of Keycloak controller

**Expected behavior**
To be able to login with my OpenShift user

--- a/address-model-lib/src/main/java/io/enmasse/address/model/KubeUtil.java
+++ b/address-model-lib/src/main/java/io/enmasse/address/model/KubeUtil.java
@@ -11,13 +11,23 @@ import io.enmasse.config.AnnotationKeys;
  */
 public class KubeUtil {
     private static int MAX_KUBE_NAME = 63 - 3; // max length of identifier - space for pod identifier
-    private static final String pattern = "[^a-z0-9\\-]";
+    private static final String addressPattern = "[^a-z0-9\\-]";
+    private static final String usernamePattern = "[^a-z0-9\\-.@_]";
+
     public static String sanitizeName(String name) {
-        if (name == null) {
+        return sanitizeWithPattern(name, addressPattern);
+    }
+
+    public static String sanitizeUserName(String name) {
+        return sanitizeWithPattern(name, usernamePattern);
+    }
+
+    private static String sanitizeWithPattern(String value, String pattern) {
+        if (value == null) {
             return null;
         }
 
-        String clean = name.toLowerCase().replaceAll(pattern, "");
+        String clean = value.toLowerCase().replaceAll(pattern, "");
         if (clean.startsWith("-")) {
             clean = clean.replaceFirst("-", "1");
         }
@@ -78,8 +88,8 @@ public class KubeUtil {
             throw new IllegalArgumentException("Name length is longer than " + MAX_KUBE_NAME + " characters");
         }
 
-        if (name.matches(pattern)) {
-            throw new IllegalArgumentException("Illegal characters found in " + name + ". Must not match " + pattern);
+        if (name.matches(addressPattern)) {
+            throw new IllegalArgumentException("Illegal characters found in " + name + ". Must not match " + addressPattern);
         }
     }
 }

--- a/keycloak-controller/src/main/java/io/enmasse/keycloak/controller/KeycloakManager.java
+++ b/keycloak-controller/src/main/java/io/enmasse/keycloak/controller/KeycloakManager.java
@@ -5,10 +5,7 @@
 
 package io.enmasse.keycloak.controller;
 
-import io.enmasse.address.model.AddressSpace;
-import io.enmasse.address.model.AuthenticationServiceType;
-import io.enmasse.address.model.EndpointSpec;
-import io.enmasse.address.model.EndpointStatus;
+import io.enmasse.address.model.*;
 import io.enmasse.config.AnnotationKeys;
 import io.enmasse.k8s.api.Watcher;
 import io.enmasse.user.api.UserApi;
@@ -113,11 +110,11 @@ public class KeycloakManager implements Watcher<AddressSpace>
                 keycloak.createRealm(addressSpace.getNamespace(), realmName, consoleUri, identityProviderParams);
                 userApi.createUser(realmName, new User.Builder()
                         .setMetadata(new UserMetadata.Builder()
-                                .setName(addressSpace.getName() + "." + userName)
+                                .setName(addressSpace.getName() + "." + KubeUtil.sanitizeUserName(userName))
                                 .setNamespace(addressSpace.getNamespace())
                                 .build())
                         .setSpec(new UserSpec.Builder()
-                                .setUsername(userName)
+                                .setUsername(KubeUtil.sanitizeUserName(userName))
                                 .setAuthentication(new UserAuthentication.Builder()
                                         .setType(UserAuthenticationType.federated)
                                         .setProvider("openshift")

--- a/user-api/src/main/java/io/enmasse/user/model/v1/UserMetadata.java
+++ b/user-api/src/main/java/io/enmasse/user/model/v1/UserMetadata.java
@@ -19,7 +19,7 @@ public class UserMetadata {
     private final String selfLink;
     private final String resourceVersion;
 
-    private final Pattern namePattern = Pattern.compile("^[a-z]+([a-z0-9\\-]*[a-z0-9]+|[a-z0-9]*)\\.[a-z0-9]+([a-z0-9\\-]*[a-z0-9]+|[a-z0-9]*)$");
+    private final Pattern namePattern = Pattern.compile("^[a-z]+([a-z0-9\\-]*[a-z0-9]+|[a-z0-9]*)\\.[a-z0-9]+([a-z0-9@.\\-]*[a-z0-9]+|[a-z0-9]*)$");
 
     @JsonCreator
     public UserMetadata(@JsonProperty("name") String name,

--- a/user-api/src/main/java/io/enmasse/user/model/v1/UserSpec.java
+++ b/user-api/src/main/java/io/enmasse/user/model/v1/UserSpec.java
@@ -17,7 +17,7 @@ public class UserSpec {
     private final String username;
     private final UserAuthentication authentication;
     private final List<UserAuthorization> authorization;
-    private final Pattern usernamePattern = Pattern.compile("^[a-z0-9]+([a-z0-9_\\-]*[a-z0-9]+|[a-z0-9]*)$");
+    private final Pattern usernamePattern = Pattern.compile("^[a-z0-9]+([a-z0-9_@.\\-]*[a-z0-9]+|[a-z0-9]*)$");
 
     @JsonCreator
     public UserSpec(@JsonProperty("username") String username,

--- a/user-api/src/test/java/io/enmasse/user/model/v1/UserModelTest.java
+++ b/user-api/src/test/java/io/enmasse/user/model/v1/UserModelTest.java
@@ -153,6 +153,16 @@ public class UserModelTest {
         createAndValidate("myspace.user_1", "user1", false);
         createAndValidate("myspace.user1", "user_1", true);
 
+        createAndValidate("myspace.user1@", "user1", false);
+        createAndValidate("myspace@.user1", "user1", false);
+        createAndValidate("@myspace.user1", "user1", false);
+        createAndValidate("myspace.@user1", "user1", false);
+        createAndValidate("myspace.user1", "user1@", false);
+        createAndValidate("myspace.user@1", "user1", true);
+        createAndValidate("myspace.user1", "user@1", true);
+
+        createAndValidate("myspace.user1@example.com", "user1@example.com", true);
+
         createAndValidate("myspace.user1-", "user1", false);
         createAndValidate("myspace-.user1", "user1", false);
         createAndValidate("-myspace.user1", "user1", false);

